llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +152,304 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+static bool isDirectReachable(SmallVector<const Value *, 5> CounterArgSet,
+                              const Value *V) {
+
+  for (auto Counter : CounterArgSet) {
+    if (V != lookThroughAIEAddressingModes(Counter))
+      return false;
+  }
+
+  return true;
+}
+
+/// This function returns some tracking information related to the upward path
+/// that starts from a Value *V to the Phi node that firstly defines a pointer
+/// value to the same ambiguous memory. The returned information can be
+/// described as follows:
+/// * int: the number of times this pointer is updated through a chain of
+///   add_2d/3d intrinsics calls. If these the pointer is updated in a non
+///   uniform way, ie. using a different addressing information (offset,
+///   (increment and size) this value will be -1. We also return {0 and {}}
+///   int case we track a value that represents the external Phi incoming
+///   value.
+/// * SmallPtrSet<Value*>: represent the constant (across updates) values
+///   that are used to update the pointer across different updates.
+/// * SmallPtrSet<Value*>: represent the chained counters across the updates.
+std::optional<IntrinsicChainInfo> trackIntrinsicChain(const Value *V) {
+  IntrinsicChainInfo ChainInfo;
+
+  if (!V->getType()->isPointerTy())
+    return std::nullopt;
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+    if (auto *GEP = dyn_cast<GEPOperator>(V)) {
+      if (GEP->hasAllZeroIndices()) {
+        // The result pointer and the first operand have the same value
+        V = GEP->getPointerOperand();
+      } else {
+        return std::nullopt; // No pointer arithmetic is allowed.
+      }
+    } else if (Operator::getOpcode(V) == Instruction::BitCast ||
+               Operator::getOpcode(V) == Instruction::AddrSpaceCast) {
+      V = cast<Operator>(V)->getOperand(0);
+      if (!V->getType()->isPointerTy())
+        return std::nullopt; // not a pointer.
+    } else {
+      if (auto *PHI = dyn_cast<PHINode>(V)) {
+        // Reached final destination.
+        return ChainInfo;
+      }
+      if (auto *AIEObject = lookThroughAIEAddressingModes(V)) {
+        IntrinsicArgsInfo NewArgSet = collectIntrinsicOperands(V);
+        // Initialize counters.
+        if (ChainInfo.IntrinsicParameters.CounterArgSet.empty()) {
+          ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
+        } else {
+          // if we already have counters, we need to reach the same
+          // AIEObject from the counters. Otherwise we can have some undesired
+          // counter update in between.
+          if (!isDirectReachable(ChainInfo.IntrinsicParameters.CounterArgSet,
+                                 AIEObject)) {
+            return std::nullopt; // not reachable.
+          }
+        }
+
+        // Update counters for the next round.
+        ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
```

Comment:
Isn't that done in the `if` above already?

Reply:
Maybe we could reverse the checks to limit the indentation, e.g.
```
        if (!ChainInfo.IntrinsicParameters.CounterArgSet.empty() &&
            !isDirectReachable(ChainInfo.IntrinsicParameters.CounterArgSet,
                                           AIEObject)) {
          // if we already have counters, we need to reach the same
          // AIEObject from the counters. Otherwise we can have some undesired
          // counter update in between.
          return std::nullopt; // not reachable.
        }
        ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
```